### ISSUE TITLE
A default app should serve i18n audio via HTTP

### ISSUE THREAD

**Issue by benlangfeld** [adhearsion/adhearsion#570]

**Comment by bklang** [comment]
If we do this, we should add a note somewhere, perhaps under a Performance section, that this should be offloaded to either direct filesystem access or a separate process.

Also, how does this work with Asterisk?

**Comment by benlangfeld** [comment]
The intention of this is for apps where FS/* do not share a filesystem with the app to work out of the box.

As yet Asterisk still does not play nicely with this idea, unlike every other IVR platform in existence :)

**Comment by nathansamson** [comment]
Hi, to be performant can we implement something like

* Ahn comes with a (simple) webserver built in (rack based?)
* The app comes with a set of audio files
* as benlangfeld mentioned, the app might not be on the same host as the VoiP platform (asterisk, freeswitch, ...)
* On startup ahn calculates the hash of the audio files in the "assets/audio" directory
* when play_asset "myfile.mp3" it will send to VOIP: http://ahn-webserver/assets/myfile-hash.mp3
* This will have a very long cache timeout,
* next time this audio file is sent the VOIP platform will (hopefully) decide it still has the file in cache, and not even do a http request, but just load from disk.

Disadvantage
* AHN needs to come with a webserver (or this can be offloaded to nginx/apache, but then the webserver config might be hard to setup, and if done wrong might not benefit)
* The AHN service needs to open an addtional port/attack vector
* We need to assume VOIP platform will come with decent caching system

Advantages
* No setup needed for users to play asset files bundled with the application, and no shared disk with the platform.
Think http://12factor.net/

I might have misunderstood the original idea, but I think this captures it?

Note: the other way should be somehow possible as well (recording an audio file, and then doing something with it on the AHN platform).

**Comment by bklang** [comment]
@nathansamson Great suggestions!  And good news:

> Ahn comes with a (simple) webserver built in (rack based?)

This is already done! See #569 

> The app comes with a set of audio files

This is also done. See #557, which includes English and Italian "Hello World" prompts

Regarding your caching questions, there is a document cache available in [Virginia](https://github.com/polysics/virginia), which was the basis for our built-in HTTP server in Adhearsion. For now we've elected to leave the document cache out, but it will likely be resurrected as a separate plugin. Check the [implementation](https://github.com/polysics/virginia/blob/develop/lib/virginia/document_cache.rb) and [documentation](https://github.com/polysics/virginia#using-the-built-in-document-cache).

However, the caching we've done so far has been for GRXML and SSML documents, since they can be expensive to generate.  We've not done caching for audio files, since they are static and don't require computation time to create. Can you go into more detail about what you are trying to solve by hashing & caching them?

**Comment by nathansamson** [comment]
The main use case seems to be where the Freeswitch platform (or Asterisk, or ...) is on a different host than the Adhearsion platform.

Imagine now AHN serving audio files via a web server. 

Call flow
* AHN (App) -> FS: Play file http://xxxx
* FS -> AHN (Web): HTTP: GET http://xxxx
* AHN (Web) -> FS: Serving the file, if a longer tune could even be megabytes.

Doing this 10s of times per second is obviously too expense for the bandwitdh usage, and also latencies to setup TCP connections might kill the peformance of the app.

Hence why Freeswitch should cache the files locally. But if you give every file a unique (hashed) name, whenever the file changes (and AHN is restarted) Freeswitch would immediately download the new file.
You can just keep the same file name, and work with Cache timeouts, or even do a HEAD request (only looking if there is an updated file without downloaded the actual file if not needed), but that will lead to either stale caches, or worse performance as the TCP connection for the HTTP call still needs to be done. And yes, you can keep the TCP connection open, and HTTP2 will even be better, but still if you don't need to access the network, then don't.

Unless I am missing something
Note that this is just a shameloss copy&paste from Rails' assets pipeline component.

And now I write this, it would even be better if the app could contain mp3 files, and then when pipelining the files could be converted to more suitable audio formats before starting to serve them. But thats another report I would say

**Comment by emcgee** [comment]
Note that using `mod_http_cache` or `mod_httapi` already include built-in caching mechanisms on the FreeSWITCH layer for audio files. Unsure if Asterisk has something similar.

**Comment by benlangfeld** [comment]
I think @nathansamson has got this dead on. Nathan, do you have any interest in contributing this feature for Adhearsion 3.0?
